### ISSUE TITLE
1.2.0: packet receive failed

### ISSUE THREAD

**Issue by jengelh** [iptraf-ng/iptraf-ng#26]
Reproducer:
* inside a systemd-234 nspawn container
* build & start 1.2.0
* select "IP traffic monitor", select `host0`
* Error ensues: `Packet receive failed / Press a key to continue`

git bisect pinpoints to this

```
043b0387fbd11670b8b09fc5bde168345648d32d is the first bad commit
commit 043b0387fbd11670b8b09fc5bde168345648d32d
Author: Vitezslav Samel <vitezslav@samel.cz>
Date:   Sun Dec 29 09:44:40 2019 +0100

    TPACKET_V3: mlock() mmap()ed address space into RAM
    
    Signed-off-by: Vitezslav Samel <vitezslav@samel.cz>

 src/capt-mmap-v3.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)
```

**Comment by pecameron** [comment]
Please verify that iptraf-ng runs as root on your system (not in a container).

iptraf-ng must run as root and must be able to set NICs into permissive mode.The only packets that can be processed are packets that reach the NIC. Containers tend to restrict activity as root.

**Comment by vsamel** [comment]
The mlock() is not strictly necessary. Try to comment it out and compile/run without it.

**Comment by jengelh** [comment]
Of course it runs as root. You don't get to the colorized menu when not doing that ("fatal: This program can be run only by the system administrator").
It's already patched out for openSUSE.

**Comment by vsamel** [comment]
> Reproducer:
>     * inside a systemd-234 nspawn container
>     * build & start 1.2.0
>     * select "IP traffic monitor", select `host0`
>     * Error ensues: `Packet receive failed / Press a key to continue`

Could you send me strace output of the failed run (I'm interested in the initialization part of capture and the failed receive part)? Something is weird because if mlock() failes then the we should not be running TPACKET_V3 nor TPACKET_V2 capturing at all but recvmmsg() or recvmsg() defaults which should work always.

**Comment by jengelh** [comment]
this gdb steprun should be more help than strace.
```
Breakpoint 1, capt_setup_mmap_v3 (capt=0x7ffc578d0eb0) at src/capt-mmap-v3.c:144
145             int version = TPACKET_V3;
(gdb) n
146             if (setsockopt(capt->fd, SOL_PACKET, PACKET_VERSION, &version, sizeof(version)) != 0)
(gdb) 
149             int hdrlen = version;
(gdb) 
150             socklen_t socklen = sizeof(hdrlen);
(gdb) 
151             if (getsockopt(capt->fd, SOL_PACKET, PACKET_HDRLEN, &hdrlen, &socklen) != 0)
(gdb) 
156             req.tp_block_nr = BLOCKS;
(gdb) 
157             req.tp_frame_nr = FRAMES;
(gdb) 
158             req.tp_frame_size = FRAME_SIZE;
(gdb) 
159             req.tp_block_size = FRAMES_PER_BLOCK * req.tp_frame_size;
(gdb) 
161             req.tp_retire_blk_tov = 20;     /* block retire timeout in msec */
(gdb) 
162             req.tp_sizeof_priv = 0;
(gdb) 
165             if(setsockopt(capt->fd, SOL_PACKET, PACKET_RX_RING, &req, sizeof(req)) != 0)
(gdb) 
168             size_t size = req.tp_block_size * req.tp_block_nr;
(gdb) 
169             void *map = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, capt->fd, 0);
(gdb) 
170             if (map == MAP_FAILED)
(gdb) 
173             if (mlock(map, size) != 0) {
(gdb) 
174                     munmap(map, size);
(gdb) p (int)errno
$1 = 12 (ENOMEM)
(gdb) n
175                     return -1;
(gdb) 
197     }
(gdb) 
capt_setup_receive_function (capt=0x7ffc578d0eb0) at src/capt.c:34
34              if (capt_setup_mmap_v3(capt) == 0)
(gdb) 
38              if (capt_setup_mmap_v2(capt) == 0)
(gdb) s
capt_setup_mmap_v2 (capt=0x7ffc578d0eb0) at src/capt-mmap-v2.c:105
105             int version = TPACKET_V2;
(gdb) n
106             if (setsockopt(capt->fd, SOL_PACKET, PACKET_VERSION, &version, sizeof(version)) != 0)
(gdb) 
107                     return -1;
(gdb) p (int)errno
$3 = 16 (EBUSY)
```

```
packet_init (pkt=0x7ffd245a3ac0) at src/packet.c:274
274             pkt->pkt_payload        = NULL;
(gdb) l
269             return PACKET_OK;
270     }
271
272     int packet_init(struct pkt_hdr *pkt)
273     {
274             pkt->pkt_payload        = NULL;
275             pkt->ethhdr             = NULL;
276             pkt->fddihdr            = NULL;
277             pkt->iphdr              = NULL;
278             pkt->ip6_hdr            = NULL;
(gdb) l
279             pkt->pkt_len            = 0;    /* signalize we have no packet prepared */
280
281             pkt->pkt_buf            = NULL;
282             pkt->from               = NULL;
283
284             return 0;       /* all O.K. */
285     }
286
287     void packet_destroy(struct pkt_hdr *pkt __unused)
288     {
(gdb) fin
Run till exit from #0  packet_init (pkt=0x7ffd245a3ac0) at src/packet.c:274
ipmon (facilitytime=0, ifptr=0x7ffd245a9d20 "host0") at src/itrafmon.c:826
826             exitloop = 0;
Value returned is $2 = 0
(gdb) n
829             clock_gettime(CLOCK_MONOTONIC, &now);
(gdb) 
830             struct timespec last_time = now;
(gdb) 
831             struct timespec next_screen_update = { 0 };
(gdb) 
832             time_t starttime = now.tv_sec;
(gdb) 
835             if (options.closedint != 0)
(gdb) 
838                     check_closed = INT_MAX;
(gdb) 
842             if (facilitytime != 0)
(gdb) 
845                     endtime = INT_MAX;
(gdb) 
847             while (!exitloop) {
(gdb) 
848                     clock_gettime(CLOCK_MONOTONIC, &now);
(gdb) 
850                     if (now.tv_sec > last_time.tv_sec) {
(gdb) 
851                             unsigned long msecs = timespec_diff_msec(&now, &last_time);
(gdb) 
853                             update_flowrates(&table, msecs);
(gdb) 
855                             print_flowrate(&table);
(gdb) 
857                             resolve_visible_entries(&table, &res);
(gdb) 
860                             printelapsedtime(now.tv_sec - starttime, 15, othptbl.borderwin);
(gdb) 
862                             print_packet_drops(capt_get_dropped(&capt), othptbl.borderwin, 40);
(gdb) 
864                             mark_timeouted_entries(&table, logging, logfile);
(gdb) 
867                             if (now.tv_sec > check_closed) {
(gdb) 
874                             if (now.tv_sec > endtime)
(gdb) 
878                             if (logging && (rotate_flag == 1)) {
(gdb) 
886                             last_time = now;
(gdb) 
890                     if (time_after(&now, &next_screen_update)) {
(gdb) 
891                             refreshtcpwin(&table, false);
(gdb) 
892                             show_stats(table.statwin, total_pkts);
(gdb) 
894                             update_panels();
(gdb) 
895                             doupdate();
(gdb) 
897                             set_next_screen_update(&next_screen_update, &now);
(gdb) 
900                     if (capt_get_packet(&capt, &pkt, &ch, table.tcpscreen) == -1) {
(gdb) s
capt_get_packet (capt=0x7ffd245a3b10, pkt=0x7ffd245a3ac0, ch=0x7ffd245a3ab0, win=0x556d029fa9d0) at src/capt.c:135
135             nfds_t nfds = 0;
(gdb) l
130     }
131
132     int capt_get_packet(struct capt *capt, struct pkt_hdr *pkt, int *ch, WINDOW *win)
133     {
134             struct pollfd pfds[2];
135             nfds_t nfds = 0;
136             int pfd_packet = -1;
137             int pfd_key = -1;
138             int ss = 0;
139             bool have_packet = capt->have_packet(capt);
(gdb) n
136             int pfd_packet = -1;
(gdb) 
137             int pfd_key = -1;
(gdb) 
138             int ss = 0;
(gdb) 
139             bool have_packet = capt->have_packet(capt);
(gdb) s
capt_have_packet_recvmmsg (capt=0x7ffd245a3b10) at src/capt-recvmmsg.c:35
35              struct capt_data_recvmmsg *data = capt->priv;
(gdb) n
37              if (capt_recvmmsg_find_filled_slot(data) != FRAMES)
(gdb) n
40                      return false;
(gdb) 
41      }
(gdb) 
capt_get_packet (capt=0x7ffd245a3b10, pkt=0x7ffd245a3ac0, ch=0x7ffd245a3ab0, win=0x556d029fa9d0) at src/capt.c:140
140             int timeout = ch ? DEFAULT_UPDATE_DELAY : -1;
(gdb) n
144             if (!have_packet) {
(gdb) 
145                     pfds[nfds].fd = capt->fd;
(gdb) 
146                     pfds[nfds].events = POLLIN;
(gdb) 
147                     pfd_packet = nfds;
(gdb) 
148                     nfds++;
(gdb) 
153             if (ch && !daemonized) {
(gdb) 
156                     clock_gettime(CLOCK_MONOTONIC, &now);
(gdb) 
159                     if (!have_packet || time_after(&now, &next_kbd_check)) {
(gdb) 
160                             pfds[nfds].fd = 0;
(gdb) 
161                             pfds[nfds].events = POLLIN;
(gdb) 
162                             pfd_key = nfds;
(gdb) 
163                             nfds++;
(gdb) 
164                             if (have_packet)
(gdb) 
167                             next_kbd_check = now;
(gdb) 
168                             time_add_msecs(&next_kbd_check, 10);
(gdb) 
172             if (nfds > 0)
(gdb) 
174                             ss = poll(pfds, nfds, timeout);
(gdb) 
175                     } while ((ss == -1) && (errno == EINTR));
(gdb) 
178             pkt->pkt_len = 0;
(gdb) p ss
$3 = 1
(gdb) n
180             if ((pfd_packet != -1) && (ss > 0)) {
(gdb) 
181                     if (pfds[pfd_packet].revents & POLLERR) {
(gdb) p pfds[pfd_packet].revents
$4 = 1
(gdb) n
186                     } else if (pfds[pfd_packet].revents & POLLIN) {
(gdb) 
188                             have_packet = true;
(gdb) 
191             if (have_packet) {
(gdb) 
192                     int ret = capt->get_packet(capt, pkt);
(gdb) s
capt_get_packet_recvmmsg (capt=0x7ffd245a3b10, pkt=0x7ffd245a3ac0) at src/capt-recvmmsg.c:45
45              struct capt_data_recvmmsg *data = capt->priv;
(gdb) n
46              int ret = 0;
(gdb) 
48              unsigned int slot = capt_recvmmsg_find_filled_slot(data);
(gdb) 
49              if (slot == FRAMES) {
(gdb) p slot
$5 = 128
(gdb) n
52                      for (unsigned int i = 0; i < FRAMES; i++) {
(gdb) l
47
48              unsigned int slot = capt_recvmmsg_find_filled_slot(data);
49              if (slot == FRAMES) {
50                      /* these are set upon return from recvmsg() so clean */
51                      /* them beforehand */
52                      for (unsigned int i = 0; i < FRAMES; i++) {
53                              data->msgvec[i].msg_hdr.msg_controllen = 0;
54                              data->msgvec[i].msg_hdr.msg_flags = 0;
55                              data->msgvec[i].msg_len = 0;
56                      }
57
58                      int received = recvmmsg(capt->fd, data->msgvec, FRAMES, MSG_TRUNC | MSG_DONTWAIT, NULL);
59                      if (received <= 0)
60                              return received;
61                      slot = 0;
62              }
63              pkt->pkt_len = data->msgvec[slot].msg_len;
64              pkt->pkt_caplen = data->msgvec[slot].msg_len;
65              if (pkt->pkt_caplen > MAX_PACKET_SIZE)
66                      pkt->pkt_caplen = MAX_PACKET_SIZE;
(gdb) b 58
Breakpoint 2 at 0x556d0256beaa: file src/capt-recvmmsg.c, line 58.
(gdb) c
Continuing.

Breakpoint 2, capt_get_packet_recvmmsg (capt=0x7ffd245a3b10, pkt=0x7ffd245a3ac0) at src/capt-recvmmsg.c:58
58                      int received = recvmmsg(capt->fd, data->msgvec, FRAMES, MSG_TRUNC | MSG_DONTWAIT, NULL);
(gdb) n
59                      if (received <= 0)
(gdb) p received
$6 = -1
(gdb) p (int)errno
$7 = 11 (EAGAIN)
(gdb) n
60                              return received;
(gdb)
75      }
(gdb)
capt_get_packet (capt=0x7ffd245a3b10, pkt=0x7ffd245a3ac0, ch=0x7ffd245a3ab0, win=0x556d029fa9d0) at src/capt.c:193
193                     if (ret <= 0)
(gdb)
194                             ss = ret;
(gdb)
197             if (ch) {
(gdb)
198                     *ch = ERR;      /* signalize we have no key ready */
(gdb)
199                     if (!daemonized && (((pfd_key != -1) && ((ss > 0) && ((pfds[pfd_key].revents & POLLIN) != 0)))))
(gdb)
203             return ss;
(gdb)
204     }
(gdb)
ipmon (facilitytime=0, ifptr=0x7ffd245a9d20 "host0") at src/itrafmon.c:900
900                     if (capt_get_packet(&capt, &pkt, &ch, table.tcpscreen) == -1) {
(gdb) p $rax
$8 = 4294967295
(gdb) n
901                             write_error("Packet receive failed");
```

**Comment by vsamel** [comment]
On Thu, Jun 11, 2020 at 01:50:39AM -0700, jengelh wrote:
> this gdb steprun should be more help than strace.

  This is superb! Found two errors. Could you please try the commits in
the fixes branch?

**Comment by jengelh** [comment]
Looks good. The
```
write_error("Locking the memory failed - continuing (suboptimally)");
```
should be dropped, because: iptraf falling back to recvmmsg is a form of "suboptimal" too, and that case does not generate a warning, so neither should a failed mlock in packetv3/v2 mode. It just adds an extra red window that the user is going to ignore anyway.

**Comment by vsamel** [comment]
Reworked. Will you try again?

**Comment by vsamel** [comment]
Fixed in #27.

**Comment by jengelh** [comment]
Works good.
